handle user updated metadata settings

kinarashah opened the pull request:
metadata settings related to k8s versions current, default and deprecated
shouldn't get overridden once user updates them

https://github.com/rancher/rancher/issues/22572

alena1108 commented:
@kinarashah would adding a test be possible? Also is there a way to restore to the default settings, once modified? If so, would be nice to add a test for it as well.

prachidamle commented:
What happens if user modifies settings and kontainer-driver-metadat gets updated to deprecate some k8s version - will that rancher setup still keep allowing the deprecated versions?

alena1108 commented:
@prachidamle it's going to be up to user/admin to be aware of that and adjust the setting. That's an implication of using a custom setting values.

Think of it similar to UI tag setting - if user adjusts it to some custom value, and then something changes in the backend code and the custom UI no longer works. Same for default docker version and other settings.

kinarashah commented:
@alena1108 @prachidamle made the following changes: 

- earlier, no default value for settings was set since they were updated dynamically. Now, on every sync, we also update default value so user can see what's the current default value based on system sync. 
- if user updates value, we validate to make sure default k8s version is present in k8s-versions-current 
- for user updated values, we make sure not to override them and use them to calculate other values like `version-to-images`, `ui-default-range` etc 
- similar to other settings in rancher, when user updates value to empty, the setting would would fallback to default value. we set the label `io.cattle.user.updated` to false 
- added tests to cover these scenarios 
- refactored a bit of code to handle these changes

prachidamle commented on pkg/controllers/management/kontainerdrivermetadata/data.go:
```diff
@@ -46,6 +41,23 @@ const (
 
+//settings corresponding to keys in setting2.MetadataSettings
+var userUpdateSettingMap = map[string]settings.Setting{
+	"k8s-version":             settings.KubernetesVersion,
```
Maybe would be good to declare these keys as const under say "rancher/pkg/settings" and reuse them across the store and kontainerdrivermetadata controller code?

prachidamle replied:
@kinarashah can we do similar change for store as well?

prachidamle commented on pkg/controllers/management/kontainerdrivermetadata/data.go:
```diff
@@ -387,97 +400,252 @@ func getWindowsName(str string) string {
+				return err
+			}
+		}
+		return nil
```
This code to update the default can be a separate method?

prachidamle commented on pkg/controllers/management/kontainerdrivermetadata/data.go:
```diff
@@ -387,97 +400,252 @@ func getWindowsName(str string) string {
-		versions = append(versions, k)
+
+	// normal setting update
+	for key, setting := range rancherUpdateSettingMap {
```
Maybe we should skip updating the settings updated by the user - by using the labels

prachidamle commented on pkg/controllers/management/kontainerdrivermetadata/data.go:
```diff
@@ -387,97 +400,252 @@ func getWindowsName(str string) string {
+		}
+	}
+
+	if err := md.updateSettingFromFields(updateSettings, true); err != nil {
```
Guess this call is to make sure defaults of all settings is updated. Do we need to update the `updateSettings` here?

prachidamle replied:
@kinarashah does it make sense to update all default values at the start of this routine and then update the setting values? Wondering if that will be good to have defaults uptodate,  in case if we hit some error updating setting values.

kinarashah replied:
don't think the order would matter either way, because we always read from value and not from default, but updated

prachidamle commented:
LGTM

alena1108 commented on pkg/controllers/management/kontainerdrivermetadata/data.go:
```diff
@@ -387,97 +401,260 @@ func getWindowsName(str string) string {
+	uiDefaultRange := fmt.Sprintf("<=%s.x", maxVersion)
+
+	return map[string]string{
+		"k8s-versions-current":            strings.Join(k8sVersionsCurrent, ","),
```
shall we put all the keys in the constants somewhere in settings? it's being referenced in settings.go already I believe
